capstone-backend/src/main/java/com/google/sps/servlets/Utility.java

```diff
@@ -36,26 +44,185 @@ public static Firestore getFirestoreDb() throws IOException {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+      return null;
+    }
+    String parameterValue = entry.getValue()[0];
+    Field[] fields = genericClass.getMyType().getDeclaredFields();
+    boolean containsParameter = false;
+    boolean parameterIsList = false;
+    for (Field field : fields) {
+      if (field.getName().equals(parameterName)) {
+        containsParameter = true;
+        if (field.getType().getName().equals("java.util.List") || field.getType().getName().equals("java.util.ArrayList")) {
+          parameterIsList = true;
+        }
+      }
+    }
+    if (!containsParameter) {
+      System.err.println("Error: The object does not have the field specified in the request parameter.");
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+      return null;
+    }
+    Query query = parameterIsList ? collectionReference.whereArrayContains(parameterName, parameterValue)
+        : collectionReference.whereEqualTo(parameterName, parameterValue);
+    while (it.hasNext()) {
+      query = addToQuery(genericClass, response, it, query);
+      if (query == null) {
+        return null;
+      }
+    }
+    return query.get();
+  }
+
+  /**
+   * Performs a get request on the given collection, allowing the requester to apply any number 
+   * of queries both for equality and to check membership in list fields. Query parameters must 
+   * match object field names exactly. If no queries are applied, the entire collection
+   * is returned. If an ID is supplied, a singleton list with that object is returned. 
+   * @param {collectionRefence} the appropriate collection of documents in the database
+   * @param {request} the request sent to the backend
+   * @param {response} the response returned from the call
+   * @param {genericClass} a Generic class, to be used in casting objects retrieved from the database
+   * @return Query the query being modified, to eventually be used to query the database.
+   */
+  public static <T> List<T> get(CollectionReference collectionReference, HttpServletRequest request, 
+      HttpServletResponse response, GenericClass<T> genericClass) throws IOException {
+    List<T> retrievedObjects = new ArrayList<>();
+    ApiFuture<QuerySnapshot> asyncQuery;
+    if (request.getParameter("id") != null) {
+      return getById(collectionReference, request, response, genericClass);
+    }
+    if (request.getParameterMap().size() == 0) {
+      asyncQuery = collectionReference.get();
+    }
+    else {
+      asyncQuery = getByField(collectionReference, request, response, genericClass);
+      if (asyncQuery == null) {
+        return null;
+      }
     }
     try {
-      Gson gson = new Gson();
-      JsonParser parser = new JsonParser();
-      JsonElement data = parser.parse(jb.toString());
-      jsonObject = data.getAsJsonObject();
+      List<QueryDocumentSnapshot> documents = asyncQuery.get().getDocuments();
```

Comment:
Not required, but I'd recommend checking out the Java [Streams API](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html) if you haven't used it before. You can use Streams `map` and `collect` methods to do this entire operation in one line.

Reply:
I'll check it out! Would be nice to clean it up.